### PR TITLE
tesseract: leptonica: add package

### PR DESCRIPTION
**Pull request by vk496**
Maintainer: me
Compile tested: ramips, R6220, 18.06.0-r7188-b0b5c64c22
Run tested: ramips, R6220, 18.06.0-r7188-b0b5c64c22, OCR image to text

Description:
Tesseract is an open source text recognizer (OCR) Engine, available
under the Apache 2.0 license. It can be used directly, or (for
programmers) using an API to extract printed text from images. It
supports a wide variety of languages.

It depends of libleptonica library

### CODE REVIEW

**Comment by cotequeiroz** [comment]
Can you split this into two commits, one for each package?

**Comment by vk496** [comment]
Done :)

File: libs/leptonica/Makefile [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#
+# Copyright (C) 2019 vk18496@gmail.com
```
Comment (diizzyy) [inline libs/leptonica/Makefile]: Use your real name or remove it

File: libs/leptonica/Makefile [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+#
+# This is free software, licensed under the GNU General Public License v3.
```
Comment (diizzyy) [inline libs/leptonica/Makefile]: Stick with v2

File: utils/tesseract/Makefile [hunk 3]
```diff
@@ -0,0 +1,231 @@
+#
+# Copyright (C) 2019 vk18496@gmail.com
```
Comment (diizzyy) [inline utils/tesseract/Makefile]: Use your real name or remove it

File: utils/tesseract/Makefile [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+#
+# This is free software, licensed under the GNU General Public License v3.
```
Comment (diizzyy) [inline utils/tesseract/Makefile]: Stick with v2

**Comment by diizzyy** [comment]
Would prefer CMake over GNU autotools for tesseract-ocr but it seems that it might be broken in 4.0 looking at upstream git log?

**Comment by BKPepe** [comment]
Could you please use for tesseract their releases, which they have tagged instead of git commit? You might use codeload. I think it will be a good idea to split Tesseract and Tessdata to separate packages as they're now in one Makefile, which is currently too long.

**Comment by vk496** [comment]
> Could you please use for tesseract their releases, which they have tagged instead of git commit? You might want to use codeload. I think it will be a good idea to split Tesseract and Tessdata to separate packages as they're now in one Makefile, which is currently too long.

How should I use codeload without external uploads?

I mean, the releases only contains automatic generated links for zip/tgz, like this one:

https://codeload.github.com/tesseract-ocr/tesseract/tar.gz/4.0.0

Since OpenWrt download script don't consider remote header name (curl `-J` argument), both tesseract-ocr and tessdata will be dowloaded as 4.0.0.tar.gz.

Is there some way to execute post-download step?

BR

**Comment by diizzyy** [comment]
https://github.com/openwrt/packages/blob/master/sound/fdk-aac/Makefile#L15

**Comment by vk496** [comment]
> https://github.com/openwrt/packages/blob/master/sound/fdk-aac/Makefile#L15

Thnks, I was missing a letter in the URL....

Only one question remain: Should I call tessdata from tesseract?

With separated packages for tesseract and tessdata, maybe would be interesting to start tessdata build process after tesseract...

**Comment by vk496** [comment]
After remove the line from both packages (tesseract and libleptonica), the package compile fine, but fails during the execution:

```bash
root@OpenWrt:~# cat test.jpg | tesseract - stdout
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
Aborted
```

I will try to remove the line ONLY from libleptonica, will see what happens....

**Comment by vk496** [comment]
I've tested all possibilities, but always get the same error if the `TARGET_LDFLAGS` is not defined. So that line is mandatory for both packages

File: utils/tessdata/Makefile [hunk 5]
```diff
@@ -0,0 +1,64 @@
+#
+# Copyright (C) 2019 Valentin Kivachuk
```
Comment (neheb) [inline utils/tessdata/Makefile]: Add email. Similar to this: https://github.com/openwrt/packages/blob/master/admin/ipmitool/Makefile#L2

File: utils/tessdata/Makefile [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+PKG_NAME:=tessdata
+PKG_VERSION:=4.0.0
+PKG_RELEASE:=1
+PKG_MAINTAINER:=Valentin Kivachuk <vk18496@gmail.com>
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://codeload.github.com/tesseract-ocr/tessdata/tar.gz/$(PKG_VERSION)?
+PKG_HASH:=5d349ac37a6ece5966bf606a6f131d628b98d88654c2f502d3c4b8bbf6ef9796
+
+
```
Comment (neheb) [inline utils/tessdata/Makefile]: put PKG_MAINTAINER here.

File: utils/tessdata/Makefile [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+endef
+
+define generate-tesseract-data-package
+  define Package/tesseract-data-$(1)
+    TITLE:=Tesseract training data for $(1) language
+    $(call Package/tesseract-data-default)
+  endef
+
+  define Package/tesseract-data-$(1)/install
+# 	$(SCRIPT_DIR)/download.pl "$(DL_DIR)" "$(1).traineddata" "skip" "" https://raw.githubusercontent.com/tesseract-ocr/tessdata/master/
```
Comment (neheb) [inline utils/tessdata/Makefile]: No idea why this is here?

File: utils/tessdata/Makefile [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	$(INSTALL_DIR) $$(1)/usr/share/tessdata
+	$(TAR) --strip=1 -C $$(1)/usr/share/tessdata/ -xvf $(DL_DIR)/$(PKG_NAME)-$(PKG_VERSION).tar.gz $(PKG_NAME)-$(PKG_VERSION)/$(1).traineddata
+  endef
+
+endef
+
+
+#$(eval $(call generate-tesseract-data-package,<lang>))
```
Comment (neheb) [inline utils/tessdata/Makefile]: Same with this

**Comment by vk496** [comment]
@neheb done.

File: utils/tesseract/Makefile [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+PKG_FIXUP:=autoreconf
+PKG_INSTALL:=1
+
+include $(INCLUDE_DIR)/package.mk
+
+TARGET_LDFLAGS += -Wl,-rpath-link=$(STAGING_DIR)/usr/lib
+
+define Package/tesseract
+  SUBMENU:=Tesseract
```
Comment (cotequeiroz) [inline utils/tesseract/Makefile]:
```suggestion
  MENU:=1
```
Remove this so it does not add a menu just for the package.  With `MENU:=1`, and removing the `SUBMENU` lines from this and tessdata, the menu will still be created, inline with the tesseract package:
```
<M> tesseract............................... Tesseract Open Source OCR Engine  --->
```
Even though I'm writing here for simplicity, you may want to move the `MENU:=1` line to the end of the block to keep a consistent style.

File: libs/leptonica/Makefile [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=http://www.leptonica.com/source/
+PKG_HASH:=e2ed2e81e7a22ddf45d2c05f0bc8b9ae7450545d995bfe28517ba408d14a5a88
+
+PKG_MAINTAINER:=Valentín Kivachuk <vk18496@gmail.com>
+PKG_LICENSE:=BSD-2-Clause
+PKG_LICENSE_FILES:=leptonica-license.txt
+
```
Comment (cotequeiroz) [inline libs/leptonica/Makefile]:
```suggestion

PKG_BUILD_PARALLEL:=1
```
Use that for faster building (tested by compiling for WRT3200ACM with `-j4`)

File: utils/tesseract/Makefile [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+PKG_RELEASE:=1
+
+PKG_SOURCE_URL:=https://codeload.github.com/tesseract-ocr/tesseract/tar.gz/$(PKG_VERSION)?
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
+PKG_HASH:=a1f5422ca49a32e5f35c54dee5112b11b99928fc9f4ee6695cdc6768d69f61dd
+
+PKG_MAINTAINER:=Valentin Kivachuk <vk18496@gmail.com>
+PKG_LICENSE:=Apache-2.0
+PKG_LICENSE_FILES:=LICENSE
+
```
Comment (cotequeiroz) [inline utils/tesseract/Makefile]:
```suggestion

PKG_BUILD_PARALLEL:=1
```
Tested by compiling for WRT3200ACM with `-j4`

File: libs/leptonica/Makefile [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+include $(INCLUDE_DIR)/package.mk
+
+define Package/libleptonica
+  SECTION:=libs
+  CATEGORY:=Libraries
+  TITLE:=A library for efficient image processing and image analysis operations
+  URL:=http://www.leptonica.com/
+  DEPENDS:=+giflib +libjpeg +libpng +libtiff +zlib
+endef
+
```
Comment (cotequeiroz) [inline libs/leptonica/Makefile]:
```suggestion

TARGET_CFLAGS:=$(filter-out -O%,$(TARGET_CFLAGS)) -O3
```

File: utils/tesseract/Makefile [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+define Package/tesseract
+  SUBMENU:=Tesseract
+  SECTION:=utils
+  CATEGORY:=Utilities
+  TITLE:=Tesseract Open Source OCR Engine
+  URL:=https://github.com/tesseract-ocr/tesseract
+  DEPENDS:=+libleptonica +libpthread +libstdcpp
+endef
+
```
Comment (cotequeiroz) [inline utils/tesseract/Makefile]:
```suggestion

TARGET_CFLAGS:=$(filter-out -O%,$(TARGET_CFLAGS)) -O3
```

**Comment by BKPepe** [comment]
ping @vk496

**Comment by vk496** [comment]
Hello,

Sorry for the delay. Holidays season :smile: 

I added all your modifications @cotequeiroz. Thank you so much for your feedback.


> Performance is poor, which makes me wonder how useful will this be in an embedded device.

In my case, was useful for automatize a daily request which involves simple captcha.

File: utils/tesseract/Makefile [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+#
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=tesseract
+PKG_VERSION:=4.0.0
+PKG_RELEASE:=1
+
+PKG_SOURCE_URL:=https://codeload.github.com/tesseract-ocr/tesseract/tar.gz/$(PKG_VERSION)?
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
```
Comment (neheb) [inline utils/tesseract/Makefile]: Place above PKG_SOURCE_URL, as elsewhere.

File: utils/tesseract/Makefile [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+PKG_LICENSE:=Apache-2.0
+PKG_LICENSE_FILES:=LICENSE
+
+PKG_BUILD_PARALLEL:=1
+PKG_FIXUP:=autoreconf
+PKG_INSTALL:=1
+
+include $(INCLUDE_DIR)/package.mk
+
+TARGET_LDFLAGS += -Wl,-rpath-link=$(STAGING_DIR)/usr/lib
```
Comment (neheb) [inline utils/tesseract/Makefile]: Is this really needed?

File: libs/leptonica/Makefile [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+define Package/libleptonica
+  SECTION:=libs
+  CATEGORY:=Libraries
+  TITLE:=A library for efficient image processing and image analysis operations
+  URL:=http://www.leptonica.org/
+  DEPENDS:=+giflib +libjpeg +libpng +libtiff +zlib
+endef
+
+TARGET_CFLAGS:=$(filter-out -O%,$(TARGET_CFLAGS)) -O3
+TARGET_LDFLAGS += -Wl,-rpath-link=$(STAGING_DIR)/usr/lib
```
Comment (neheb) [inline libs/leptonica/Makefile]: Is this really needed?

File: utils/tesseract/Makefile (hunk 15 above)
Comment (vk496) [reply utils/tesseract/Makefile]: Yes. I answered this this question before. If I omit the LDFLAGS, the final binary will be broken (segfaults after try to execute it)

File: libs/leptonica/Makefile (hunk 16 above)
Comment (vk496) [reply libs/leptonica/Makefile]: Yes, exactly the same problem with LDFLAGS (+ @cotequeiroz addition to improve times)

**Pull request merged**